x/monitoringc/keeper/invariants.go

```diff
@@ -0,0 +1,64 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/tendermint/spn/x/monitoringc/types"
+)
+
+const (
+	missingVerifiedClientIDRoute = "missing-verified-client-id"
+)
+
+// RegisterInvariants registers all module invariants
+func RegisterInvariants(ir sdk.InvariantRegistry, k Keeper) {
+	ir.RegisterRoute(types.ModuleName, missingVerifiedClientIDRoute,
+		MissingVerifiedClientIDInvariant(k))
+}
+
+// AllInvariants runs all invariants of the module.
+func AllInvariants(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		return MissingVerifiedClientIDInvariant(k)(ctx)
+	}
+}
+
+// MissingVerifiedClientIDInvariant checks if any of the clientIDs in `LaunchIDFromVerifiedClientID` or `ProviderClientID`
+// does not have a corresponding entry in `VerifiedClientID`
+func MissingVerifiedClientIDInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		allVerifiedClientID := k.GetAllVerifiedClientID(ctx)
+		allLaunchIDFromVerifiedClientlID := k.GetAllLaunchIDFromVerifiedClientID(ctx)
+		allProviderClientID := k.GetAllProviderClientID(ctx)
+		clientIDMap := make(map[string]bool)
+		for _, verifiedClientID := range allVerifiedClientID {
+			for _, clientID := range verifiedClientID.ClientIDs {
+				clientIDMap[clientIDKey(verifiedClientID.LaunchID, clientID)] = true
+			}
+		}
+		for _, providerClientID := range allProviderClientID {
+			if _, ok := clientIDMap[clientIDKey(providerClientID.LaunchID, providerClientID.ClientID)]; !ok {
+				return sdk.FormatInvariant(
+					types.ModuleName, missingVerifiedClientIDRoute,
+					"client id from providerClientID list not found",
+				), true
+			}
+		}
+		for _, launchIDFromVerifiedClientID := range allLaunchIDFromVerifiedClientlID {
+			if _, ok := clientIDMap[clientIDKey(launchIDFromVerifiedClientID.LaunchID, launchIDFromVerifiedClientID.ClientID)]; !ok {
+				return sdk.FormatInvariant(
+					types.ModuleName, missingVerifiedClientIDRoute,
+					"client id from launchIDFromVerifiedClientID list not found",
+				), true
+			}
+		}
```

Comment:
The `launchIDFromVerifiedClientID` entry is created  when the associated when the client ID is added to `VerifiedClientID`. So that when the IBC connection is performed the corresponding launch ID can be retrieved.

Under some condition (`MsgRevertLaunch`) the `VerifiedClientID` is erased to prevent make an IBC connection with the chain but the `launchIDFromVerifiedClientID` is not removed, which would break this invariant.

I think the condition should be the opposite: Every verified client ID should have an associated `launchIDFromVerifiedClientID` to allow the correct configuration once the connection is performed.

A `launchIDFromVerifiedClientID` without an actual verified client ID would be just dead data in the store

Reply:
Sounds good. I've removed completely the check for `ProviderClientID` and implemented the invariant how you suggested